File: java/yass/test/ch/softappeal/yass/serialize/test/FastSerializerTest.java
package ch.softappeal.yass.serialize.test;

import ch.softappeal.yass.serialize.Reader;
import ch.softappeal.yass.serialize.Serializer;
import ch.softappeal.yass.serialize.Writer;
import ch.softappeal.yass.serialize.contract.C1;
import ch.softappeal.yass.serialize.contract.C2;
import ch.softappeal.yass.serialize.contract.Color;
import ch.softappeal.yass.serialize.contract.E1;
import ch.softappeal.yass.serialize.contract.E2;
import ch.softappeal.yass.serialize.fast.BaseTypeHandler;
import ch.softappeal.yass.serialize.fast.BaseTypeHandlers;
import ch.softappeal.yass.serialize.fast.FastSerializer;
import ch.softappeal.yass.serialize.fast.SimpleFastSerializer;
import ch.softappeal.yass.serialize.fast.TaggedFastSerializer;
import ch.softappeal.yass.serialize.fast.TypeDesc;
import ch.softappeal.yass.util.Nullable;
import ch.softappeal.yass.util.Tag;
import ch.softappeal.yass.util.TestUtils;
import org.junit.Assert;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.PrintWriter;
import java.util.List;

public class FastSerializerTest {

    @Test public void baseEnumeration() {
        try {
            new TaggedFastSerializer(
                List.of(new TypeDesc(1, new BaseTypeHandler<>(Color.class) {
                    @Override public Color read(final Reader reader) {
                        return null;
                    }
                    @Override public void write(final Color value, final Writer writer) {
                        // empty
                    }
                })),
                List.of()
            );
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("base type 'ch.softappeal.yass.serialize.contract.Color' is an enumeration", e.getMessage());
        }
    }

    @Test public void duplicatedClass() {
        try {
            new TaggedFastSerializer(List.of(), List.of(Color.class, Color.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("type 'ch.softappeal.yass.serialize.contract.Color' already added", e.getMessage());
        }
    }

    @Test public void abstractClass() {
        try {
            new SimpleFastSerializer(List.of(), List.of(FastSerializer.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("type 'ch.softappeal.yass.serialize.fast.FastSerializer' is abstract", e.getMessage());
        }
    }

    @Test public void illegalInterface() {
        try {
            new SimpleFastSerializer(List.of(), List.of(AutoCloseable.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("type 'java.lang.AutoCloseable' is abstract", e.getMessage());
        }
    }

    @Test public void illegalAnnotation() {
        try {
            new SimpleFastSerializer(List.of(), List.of(Nullable.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("type 'ch.softappeal.yass.util.Nullable' is abstract", e.getMessage());
        }
    }

    @Test public void illegalEnumeration() {
        try {
            new SimpleFastSerializer(List.of(), List.of(), List.of(Color.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("type 'ch.softappeal.yass.serialize.contract.Color' is an enumeration", e.getMessage());
        }
    }

    @Test public void missingType() throws Exception {
        final Serializer serializer = new TaggedFastSerializer(List.of(), List.of());
        try {
            JavaSerializerTest.copy(serializer, Color.BLUE);
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("missing type 'ch.softappeal.yass.serialize.contract.Color'", e.getMessage());
        }
    }

    @Test public void taggedPrint() {
        TestUtils.compareFile("ch/softappeal/yass/serialize/test/TaggedFastSerializerTest.numbers.txt", SerializerTest.TAGGED_FAST_SERIALIZER::print);
    }

    @Test public void simplePrint() {
        TestUtils.compareFile("ch/softappeal/yass/serialize/test/SimpleFastSerializerTest.numbers.txt", SerializerTest.SIMPLE_FAST_SERIALIZER::print);
    }

    @Test public void bytes() throws Exception {
        TestUtils.compareFile("ch/softappeal/yass/serialize/test/TaggedFastSerializerTest.bytes.txt", new TestUtils.Printer() {
            void write(final PrintWriter printer, final Object value) throws Exception {
                final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
                SerializerTest.TAGGED_FAST_SERIALIZER.write(value, Writer.create(buffer));
                final byte[] bytes = buffer.toByteArray();
                for (final byte b : bytes) {
                    printer.print(" " + b);
                }
                printer.println();
            }
            @Override public void print(final PrintWriter printer) throws Exception {
                write(printer, SerializerTest.createGraph());
                write(printer, SerializerTest.createNulls());
                write(printer, SerializerTest.createValues());
            }
        });
    }

    private static final Serializer V1_SERIALIZER = new TaggedFastSerializer(
        List.of(new TypeDesc(3, BaseTypeHandlers.INTEGER)), List.of(E1.class, C1.class)
    );
    private static final Serializer V2_SERIALIZER = new TaggedFastSerializer(
        List.of(new TypeDesc(3, BaseTypeHandlers.INTEGER)), List.of(E2.class, C2.class)
    );

    private static Object copy(final Object input) throws Exception {
        final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        final Writer writer = Writer.create(buffer);
        V1_SERIALIZER.write(input, writer);
        writer.writeByte((byte)123); // write sentinel
        final Reader reader = Reader.create(new ByteArrayInputStream(buffer.toByteArray()));
        final Object output = V2_SERIALIZER.read(reader);
        Assert.assertTrue(reader.readByte() == 123); // check sentinel
        return output;
    }

    @Test public void versioning() throws Exception {
        final C2 c2 = (C2)copy(new C1(42));
        Assert.assertTrue(c2.i1 == 42);
        Assert.assertNull(c2.i2);
        Assert.assertTrue(c2.i2() == 13);
        Assert.assertSame(copy(E1.c1), E2.c1);
        Assert.assertSame(copy(E1.c2), E2.c2);
    }

    public static class MissingClassTag {
        @Tag(1) int i;
    }

    @Test public void missingClassTag() {
        try {
            new TaggedFastSerializer(List.of(), List.of(MissingClassTag.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("missing tag for 'class ch.softappeal.yass.serialize.test.FastSerializerTest$MissingClassTag'", e.getMessage());
        }
    }

    @Test public void duplicatedTypeTag() {
        try {
            new TaggedFastSerializer(List.of(), List.of(C1.class, C2.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            System.out.println(e.getMessage());
        }
    }

    @Tag(-1) public static class InvalidTypeTag {
        // empty
    }

    @Test public void invalidTypeTag() {
        try {
            new TaggedFastSerializer(List.of(), List.of(InvalidTypeTag.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals("id -1 for type 'ch.softappeal.yass.serialize.test.FastSerializerTest.InvalidTypeTag' must be >= 0", e.getMessage());
        }
    }

    @Tag(0) public static class InvalidFieldTag {
        @Tag(0) int i;
    }

    @Test public void invalidFieldTag() {
        try {
            new TaggedFastSerializer(List.of(), List.of(InvalidFieldTag.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            Assert.assertEquals(
                "id 0 for field 'int ch.softappeal.yass.serialize.test.FastSerializerTest$InvalidFieldTag.i' must be >= 1",
                e.getMessage()
            );
        }
    }

    @Tag(0) public static class DuplicatedFieldTag {
        @Tag(1) int i1;
        @Tag(1) int i2;
    }

    @Test public void duplicatedFieldTag() {
        try {
            new TaggedFastSerializer(List.of(), List.of(DuplicatedFieldTag.class));
            Assert.fail();
        } catch (final IllegalArgumentException e) {
            System.out.println(e.getMessage());
        }
    }

}

File: kotlin/yass/test/ch/softappeal/yass/serialize/FastSerializerTest.kt
package ch.softappeal.yass.serialize

import ch.softappeal.yass.Tag
import ch.softappeal.yass.compareFile
import ch.softappeal.yass.serialize.fast.BaseTypeSerializer
import ch.softappeal.yass.serialize.fast.FastSerializer
import ch.softappeal.yass.serialize.fast.IntSerializer
import ch.softappeal.yass.serialize.fast.TypeDesc
import ch.softappeal.yass.serialize.fast.print
import ch.softappeal.yass.serialize.fast.simpleFastSerializer
import ch.softappeal.yass.serialize.fast.taggedFastSerializer
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.PrintWriter
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertSame
import kotlin.test.assertTrue
import kotlin.test.fail

class FastSerializerTest {

    class A(val a: Int)

    @Test
    fun duplicatedField() = try {
        val field = A::class.java.getDeclaredField("a")
        object : FastSerializer() {
            init {
                addClass(999, String::class.java, false, mapOf(1 to field, 2 to field))
            }
        }
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals(
            "duplicated field name 'private final int ch.softappeal.yass.serialize.FastSerializerTest${'$'}A.a' and 'private final int ch.softappeal.yass.serialize.FastSerializerTest${'$'}A.a' not allowed in class hierarchy",
            e.message
        )
    }

    @Test
    fun notAnEnumeration() = try {
        object : FastSerializer() {
            init {
                addEnum(999, String::class.java)
            }
        }
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'java.lang.String' is not an enumeration", e.message)
    }

    @Test
    fun baseEnumeration() = try {
        taggedFastSerializer(
            listOf(TypeDesc(1, object : BaseTypeSerializer<Color>(Color::class.java) {
                override fun read(reader: Reader) = Color.BLUE
                override fun write(writer: Writer, value: Color) {}
            })),
            listOf()
        )
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("base type 'ch.softappeal.yass.serialize.Color' is an enumeration", e.message)
    }

    @Test
    fun duplicatedClass() = try {
        taggedFastSerializer(listOf(), listOf(Color::class.java, Color::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'ch.softappeal.yass.serialize.Color' already added", e.message)
    }

    @Test
    fun abstractClass() = try {
        simpleFastSerializer(listOf(), listOf(FastSerializer::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'ch.softappeal.yass.serialize.fast.FastSerializer' is abstract", e.message)
    }

    @Test
    fun illegalInterface() = try {
        simpleFastSerializer(listOf(), listOf(AutoCloseable::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'java.lang.AutoCloseable' is abstract", e.message)
    }

    @Test
    fun illegalAnnotation() = try {
        simpleFastSerializer(listOf(), listOf(Tag::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'ch.softappeal.yass.Tag' is abstract", e.message)
    }

    @Test
    fun illegalEnumeration() = try {
        simpleFastSerializer(listOf(), listOf(), listOf(Color::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("type 'ch.softappeal.yass.serialize.Color' is an enumeration", e.message)
    }

    @Test
    fun missingType() {
        val serializer = taggedFastSerializer(listOf(), listOf())
        try {
            copy(serializer, Color.BLUE)
            fail()
        } catch (e: IllegalStateException) {
            assertEquals("missing type 'ch.softappeal.yass.serialize.Color'", e.message)
        }

    }

    class MissingClassTag {
        @Tag(1)
        internal var i: Int = 0
    }

    @Test
    fun missingClassTag() = try {
        taggedFastSerializer(listOf(), listOf(MissingClassTag::class.java))
        fail()
    } catch (e: IllegalStateException) {
        assertEquals("missing tag for 'class ch.softappeal.yass.serialize.FastSerializerTest${'$'}MissingClassTag'", e.message)
    }

    @Test
    fun duplicatedTypeTag() = try {
        taggedFastSerializer(listOf(), listOf(C1::class.java, C2::class.java))
        fail()
    } catch (e: IllegalStateException) {
        assertEquals(
            "type id 120 used for 'ch.softappeal.yass.serialize.C2' and 'ch.softappeal.yass.serialize.C1'",
            e.message
        )
    }

    @Tag(-1)
    class InvalidTypeTag// empty

    @Test
    fun invalidTypeTag() = try {
        taggedFastSerializer(listOf(), listOf(InvalidTypeTag::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals("id -1 for type 'ch.softappeal.yass.serialize.FastSerializerTest.InvalidTypeTag' must be >= 0", e.message)
    }

    @Tag(0)
    class InvalidFieldTag {
        @Tag(0)
        internal var i: Int = 0
    }

    @Test
    fun invalidFieldTag() = try {
        taggedFastSerializer(listOf(), listOf(InvalidFieldTag::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals(
            "id 0 for field 'private int ch.softappeal.yass.serialize.FastSerializerTest${'$'}InvalidFieldTag.i' must be >= 1",
            e.message
        )
    }

    @Tag(0)
    class DuplicatedFieldTag {
        @Tag(1)
        internal var i1: Int = 0
        @Tag(1)
        internal var i2: Int = 0
    }

    @Test
    fun duplicatedFieldTag() = try {
        taggedFastSerializer(listOf(), listOf(DuplicatedFieldTag::class.java))
        fail()
    } catch (e: IllegalArgumentException) {
        assertEquals(
            "tag 1 used for fields 'private int ch.softappeal.yass.serialize.FastSerializerTest${'$'}DuplicatedFieldTag.i2' and 'private int ch.softappeal.yass.serialize.FastSerializerTest${'$'}DuplicatedFieldTag.i1'",
            e.message
        )
    }

    @Test
    fun taggedPrint() {
        compareFile("ch/softappeal/yass/serialize/TaggedFastSerializerTest.numbers.txt") { TAGGED_FAST_SERIALIZER.print(it) }
    }

    @Test
    fun simplePrint() {
        compareFile("ch/softappeal/yass/serialize/SimpleFastSerializerTest.numbers.txt") { SIMPLE_FAST_SERIALIZER.print(it) }
    }

    @Test
    fun bytes() {
        compareFile("ch/softappeal/yass/serialize/TaggedFastSerializerTest.bytes.txt") { printer ->
            fun write(printer: PrintWriter, value: Any) {
                val buffer = ByteArrayOutputStream()
                TAGGED_FAST_SERIALIZER.write(writer(buffer), value)
                val bytes = buffer.toByteArray()
                for (b in bytes) {
                    printer.print(" $b")
                }
                printer.println()
            }
            write(printer, createGraph())
            write(printer, createNulls())
            write(printer, createValues())
        }
    }

    @Test
    fun versioning() {
        val v1serializer = taggedFastSerializer(listOf(TypeDesc(3, IntSerializer)), listOf(E1::class.java, C1::class.java))
        val v2serializer = taggedFastSerializer(listOf(TypeDesc(3, IntSerializer)), listOf(E2::class.java, C2::class.java))
        fun copy(input: Any): Any {
            val buffer = ByteArrayOutputStream()
            val writer = writer(buffer)
            v1serializer.write(writer, input)
            writer.writeByte(123.toByte()) // write sentinel
            val reader = reader(ByteArrayInputStream(buffer.toByteArray()))
            val output = v2serializer.read(reader)
            assertTrue(reader.readByte() == 123.toByte()) // check sentinel
            return output!!
        }

        val c2 = copy(C1(42)) as C2
        assertTrue(c2.i1 == 42)
        assertNull(c2.i2)
        assertTrue(c2.i2() == 13)
        assertSame(copy(E1.c1), E2.c1)
        assertSame(copy(E1.c2), E2.c2)
    }

}
